erigon-lib/p2p/sentry/sentrymultiplexer.go

```diff
@@ -133,84 +136,138 @@ func (m *sentryMultiplexer) HandShake(ctx context.Context, in *emptypb.Empty, op
 	}
 
-	err := g.Wait()
+	peerReplies, err := m.peersByClient(ctx, minProtocol, opts...)
 
 	if err != nil {
 		return nil, err
 	}
 
-	return &sentryproto.SentPeers{Peers: allSentPeers}, nil
-}
+	clientIndex := -1
+	peerId := AsPeerIdString(in.PeerId)
+
+CLIENTS:
+	for i, reply := range peerReplies {
+		for _, peer := range reply.GetPeers() {
+			if peer.Id == peerId {
+				clientIndex = i
+				break CLIENTS
+			}
+		}
+	}
+
+	if clientIndex < 0 {
+		return nil, fmt.Errorf("peer not found: %s", peerId)
+	}
 
-func (m *sentryMultiplexer) SendMessageById(ctx context.Context, in *sentryproto.SendMessageByIdRequest, opts ...grpc.CallOption) (*sentryproto.SentPeers, error) {
 	g, gctx := errgroup.WithContext(ctx)
 
 	var allSentPeers []*typesproto.H512
-	var allSentMutex sync.RWMutex
 
-	for _, client := range m.clients {
-		client := client
+	g.Go(func() error {
+		sentPeers, err := m.clients[clientIndex].SendMessageById(gctx, in, opts...)
 
-		g.Go(func() error {
-			sentPeers, err := client.SendMessageById(gctx, in, opts...)
+		if err != nil {
+			return err
+		}
 
-			if err != nil {
-				return err
-			}
+		allSentPeers = sentPeers.GetPeers()
 
-			allSentMutex.Lock()
-			defer allSentMutex.Unlock()
+		return nil
+	})
 
-			allSentPeers = append(allSentPeers, sentPeers.GetPeers()...)
+	if err = g.Wait(); err != nil {
+		return nil, err
+	}
 
-			return nil
-		})
+	return &sentryproto.SentPeers{Peers: allSentPeers}, nil
+}
+
+func (m *sentryMultiplexer) SendMessageToRandomPeers(ctx context.Context, in *sentryproto.SendMessageToRandomPeersRequest, opts ...grpc.CallOption) (*sentryproto.SentPeers, error) {
+	minProtocol := MinProtocol(in.Data.Id)
+
+	if minProtocol < 0 {
+		return nil, fmt.Errorf("unknown protocol for: %s", in.Data.Id.String())
 	}
 
-	err := g.Wait()
+	peerReplies, err := m.peersByClient(ctx, minProtocol, opts...)
 
 	if err != nil {
 		return nil, err
 	}
 
-	return &sentryproto.SentPeers{Peers: allSentPeers}, nil
-}
+	type peer struct {
+		clientIndex int
+		peerId      *typesproto.H512
+	}
+
+	peers := map[string]peer{}
+
+	for i, reply := range peerReplies {
+		for _, p := range reply.GetPeers() {
+			if _, ok := peers[p.Id]; !ok {
+				peers[p.Id] = peer{
+					clientIndex: i,
+					peerId:      gointerfaces.ConvertHashToH512([64]byte(common.Hex2Bytes(p.Id))),
+				}
+			}
+		}
+	}
 
-func (m *sentryMultiplexer) SendMessageToRandomPeers(ctx context.Context, in *sentryproto.SendMessageToRandomPeersRequest, opts ...grpc.CallOption) (*sentryproto.SentPeers, error) {
 	g, gctx := errgroup.WithContext(ctx)
 
 	var allSentPeers []*typesproto.H512
 	var allSentMutex sync.RWMutex
+	var sentCount uint64
 
-	for _, client := range m.clients {
-		client := client
+	for _, peer := range peers {
+		peer := peer
 
 		g.Go(func() error {
-			sentPeers, err := client.SendMessageToRandomPeers(gctx, in, opts...)
+			sentPeers, err := m.clients[peer.clientIndex].SendMessageById(gctx, &sentryproto.SendMessageByIdRequest{
```

Comment:
It seems like we are not preserving the meaning of `SendMessageToRandomPeers` and are making it non random. We probably should do a shuffle of the peers list if their size is > MaxPeers in the multiplexer before this for loop to preserve it